swap/htlc.go

```
@@ -515,3 +559,239 @@ func (h *HtlcScriptV2) MaxTimeoutWitnessSize() int {
 func (h *HtlcScriptV2) SuccessSequence() uint32 {
 	return 1
 }
+
+// HtlcScriptV3 encapsulates the htlc v3 script.
+type HtlcScriptV3 struct {
+	// The final locking script for the timeout path which is available to
+	// the sender after the set blockheight.
+	TimeoutScript []byte
+
+	// The final locking script for the success path in which the receiver
+	// reveals the preimage.
+	SuccessScript []byte
+
+	// The public key for the keyspend path which bypasses the above two
+	// locking scripts.
+	InternalPubKey *btcec.PublicKey
+
+	// The taproot public key which is created with the above 3 inputs.
+	TaprootKey *btcec.PublicKey
+}
+
+// newHTLCScriptV3 constructs a HtlcScipt with the HTLC V3 taproot script.
+func newHTLCScriptV3(cltvExpiry int32, senderHtlcKey,
+	receiverHtlcKey [33]byte, sharedKey *btcec.PublicKey,
+	swapHash lntypes.Hash) (*HtlcScriptV3, error) {
+
+	receiverPubKey, err := btcec.ParsePubKey(
```

We should return an error if the passed `sharedKey == nil`.